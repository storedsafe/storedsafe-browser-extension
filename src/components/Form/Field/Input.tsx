import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Input.scss';

type InputProps = Pick<FieldProps, 'name' | 'type' | 'value' | 'onChange' | 'onBlur' | 'label'>;
export const Input: React.SFC<InputProps | FieldAttributes> = ({
  name,
  type,
  value,
  label,
  onChange,
  onBlur,
  ...attributes
}: InputProps) => (
  <div className={`field field-input field-input-${type}`}>
    <label htmlFor={name}>
      <span className="field-label">{label}</span>
      <div className="field-container">
        <input
          className="field-control"
          id={name}
          name={name}
          type={type}
          value={['number', 'range'].includes(type) && isNaN(value as number) ? '' : value as string | number}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onChange && onChange(target.valueAsNumber, name);
            } else {
              onChange && onChange(target.value, name);
            }
          }}
          onBlur={({ target }: React.FocusEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onBlur && onBlur(target.valueAsNumber, name);
            } else {
              onBlur && onBlur(target.value, name);
            }
          }}
          {...attributes}
        />
      </div>
    </label>
  </div>
);

Input.defaultProps = {
  type: 'text',
  value: '',
}
