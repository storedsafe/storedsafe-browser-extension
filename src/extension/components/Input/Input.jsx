/* eslint react/jsx-props-no-spreading: 0 */
import React from 'react';
import PropTypes from 'prop-types';

function Input({
  title,
  type,
  name,
  children,
  ...attrs
}) {
  if (type === 'checkbox') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <input
          type={type}
          name={name}
          id={name}
          {...attrs}
        />
        <span className="checkmark" />
      </label>
    );
  }

  if (type === 'select') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <div>
          <select
            name={name}
            id={name}
            {...attrs}
          >
            {children}
          </select>
        </div>
      </label>
    );
  }

  if (type === 'button' || type === 'submit') {
    return (
      <input
        className={type}
        type={type}
        name={name}
        id={name}
        value={title}
        {...attrs}
      />
    );
  }

  if (type === 'textarea') {
    return (
      <label htmlFor={name} className={type}>
        <span>{title}</span>
        <textarea
          name={name}
          id={name}
          {...attrs}
        />
      </label>
    );
  }

  return (
    <label htmlFor={name} className={type}>
      <span>{title}</span>
      <input
        type={type}
        name={name}
        id={name}
        {...attrs}
      />
    </label>
  );
}

Input.defaultProps = {
  children: null,
};

Input.propTypes = {
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  children: PropTypes.node,
};

export default Input;
