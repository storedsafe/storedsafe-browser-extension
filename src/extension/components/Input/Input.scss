@use '../../base.scss';

/* Flow Styling */
label,
input[type='submit'],
input[type='button'] {
  margin-bottom: 1em;
}

label {
  display: flex;
  flex-direction: column;
  width: 100%;

  span {
    font-family: 'Montserrat', Helvetica, sans-serif;
    text-transform: uppercase;
  }
}

/* Component Styling */
input,
button,
select {
  border: 0;
  box-sizing: border-box;
  outline: 0;
  padding: 1em;
}

input:not([type=submit]):not([type=button]):not([type=checkbox]):not([type=radio]),
select,
option {
  background-color: base.$bg-light;
  border-bottom: 1px solid base.$border;
  color: #000;
  font-family: sans-serif;
  font-size: 1em;
  transition: base.$animation-duration ease;

  &:hover,
  &:focus {
    border-bottom-color: base.$accent;
  }

  &:focus {
    background-color: base.$bg-medium;
  }
}

option {
  // Chrome option inherits from select, firefox doesn't
  background-color: base.$bg-light;
}

button,
input[type=button],
input[type=submit] {
  background: base.$primary;
  border: 1px solid base.$bg-light;
  border-radius: base.$border-radius;
  color: base.$fg-light;
  font-weight: bold;
  text-transform: uppercase;
  transition: base.$animation-duration ease;
  width: 100%;

  &:hover,
  &:focus,
  &:active {
    background-color: base.$bg-light;
    border: 1px solid base.$accent;
    color: base.$accent;
  }

  &:active {
    background-color: base.$bg-medium;
  }
}

/* Disable and replace default style */
.checkbox {
  align-items: center;
  cursor: pointer;
  flex-direction: row-reverse;
  justify-content: flex-end;

  // Disable default checkbox
  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  // Create custom checkbox
  .checkmark {
    border: 1px solid base.$border;
    border-radius: base.$border-radius;
    height: 1.5em;
    margin-right: 0.5em;
    transition: base.$animation-duration ease;
    width: 1.5em;
  }

  // Create checkmark
  .checkmark::after {
    border: solid base.$fg-light;
    border-width: 0 0.2em 0.2em 0;
    box-sizing: border-box;
    // Ensure consistent size across browsers
    // Firefox acts weirdly if using em units
    font-size: 1rem;
    height: 60%;
    // Top margin less to prefer being pushed up rather than
    // down with pixel errors
    margin: 19% 35%;
    transform: rotate(45deg);
    width: 30%;
  }

  // Hide checkmark if not checked
  .checkmark::after {
    content: '';
    display: none;
  }

  // Show checkmark if checked
  input:checked ~ .checkmark::after {
    display: block;
  }

  // General appearance based on state
  &:hover input ~ .checkmark {
    border: 1px solid base.$accent;
  }

  input:focus ~ .checkmark {
    background-color: base.$bg-medium;
    border-color: base.$accent;
  }

  &:active input ~ .checkmark {
    background-color: base.$bg-medium;
  }

  input:checked ~ .checkmark {
    background-color: base.$accent;
  }
}

select {
  -moz-appearance: none; // stylelint-disable-line
  -webkit-appearance: none; // stylelint-disable-line
  appearance: none;
  padding-right: 2.25em;
  width: 100%;
}

.select div {
  position: relative;

  &::after {
    border: solid transparent;
    border-radius: base.$border-radius;
    border-top-color: base.$bg-dark;
    border-width: 0.7em 0.5em;
    box-sizing: border-box;
    content: '';
    height: 1em;
    margin-top: -0.35em;
    position: absolute;
    right: 0.75em;
    top: 50%;
    transition: base.$animation-duration;
    width: 1em;
  }
}
