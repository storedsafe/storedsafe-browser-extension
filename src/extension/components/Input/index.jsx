import Input from './Input';
import useForm from './InputHooks';
import './Input.scss';

export { useForm };
export default Input;
