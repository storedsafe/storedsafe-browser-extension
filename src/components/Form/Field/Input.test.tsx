import * as React from 'react';
import { Input } from './Input';
import { shallow, mount } from 'enzyme';

const onChange = jest.fn();
const onBlur = jest.fn();

beforeEach(() => {
  onChange.mockClear();
  onBlur.mockClear();
});

test('<Input />', () => {
  const wrapper = shallow(<Input name="test" label="Test" />);
  const input = wrapper.find('input#test');
  expect(input.props().type).toBe('text');
  expect(wrapper).toMatchSnapshot();
});

test('<Input password />', () => {
  const wrapper = shallow(<Input name="test" label="Test" type="password" />);
  const input = wrapper.find('input#test');
  expect(input.props().type).toBe('password');
  expect(wrapper).toMatchSnapshot();
});

test('<Input number />, NaN', () => {
  const wrapper = shallow(<Input name="test" label="Test" type="number" value={NaN} />);
  const input = wrapper.find('input#test');
  expect(input.props().value).toBe('');
  expect(wrapper).toMatchSnapshot();
});

test('<Input />, change', () => {
  const wrapper = mount(<Input name="test" label="Test" value="foo" onChange={onChange} />);
  wrapper.find('input#test').simulate('change', { target: { name: 'test', value: 'bar' } });
  expect(onChange).toHaveBeenCalledWith('bar', 'test');
});

test('<Input number />, change', () => {
  const wrapper = mount(<Input name="test" label="Test" type="number" value={1} onChange={onChange} />);
  wrapper.find('input#test').simulate('change', { target: { name: 'test', valueAsNumber: 2 } });
  expect(onChange).toHaveBeenCalledWith(2, 'test');
});

test('<Input />, blur', () => {
  const wrapper = mount(<Input name="test" label="Test" type="text" value="foo" onBlur={onBlur} />);
  wrapper.find('input#test').simulate('blur', { target: { name: 'test', value: 'bar' } });
  expect(onBlur).toHaveBeenCalledWith('bar', 'test');
});

test('<Input number />, blur', () => {
  const wrapper = mount(<Input name="test" label="Test" type="number" value={1} onBlur={onBlur} />);
  wrapper.find('input#test').simulate('blur', { target: { name: 'test', valueAsNumber: 2 } });
  expect(onBlur).toHaveBeenCalledWith(2, 'test');
});
