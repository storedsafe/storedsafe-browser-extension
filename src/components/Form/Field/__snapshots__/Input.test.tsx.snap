// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> 1`] = `ShallowWrapper {}`;

exports[`<Input number />, NaN 1`] = `ShallowWrapper {}`;

exports[`<Input password /> 1`] = `ShallowWrapper {}`;
